### PR TITLE
106176 validations on queries vol2

### PR DESCRIPTION
**Pull request by janid1967**
[AB#106176](https://statoildeveloper.visualstudio.com/b6d97460-2e53-47ae-aad4-b41396d90828/_workitems/edit/106176) 

This PR has many file changed since Validator interfaces are moved from Command to Domain project, and implementation of these are moved from Command to WebApi project. Unit tests for such are also moved. 

The goal for PR is to be able to validate Queries (as for Commands) to be able to return NotFound (404) to client when client getting unknown/non-existing stuff. When performing query validation, the IPipelineBehavior CheckValidProjectBehavior became superfluous (now removed)

### CODE REVIEW

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemAttachmentDownloadUrl/GetPunchItemAttachmentDownloadUrlQueryValidator.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        IAttachmentService attachmentService)
+    {
+        RuleLevelCascadeMode = CascadeMode.Stop;
+        ClassLevelCascadeMode = CascadeMode.Stop;
+
+        RuleFor(query => query)
+            .MustAsync((query, cancellationToken) => BeAnExistingPunchItemAsync(query.PunchItemGuid, cancellationToken))
+            .WithMessage(query => $"Punch item with this guid does not exist! Guid={query.PunchItemGuid}")
+            .WithState(_ => new EntityNotFoundException())
+            .MustAsync((command, cancellationToken) => BeAnExistingAttachment(command.AttachmentGuid, cancellationToken))
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemAttachmentDownloadUrl/GetPunchItemAttachmentDownloadUrlQueryValidator.cs:24]: command -> query?

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemsInProject/GetPunchItemsInProjectQueryValidator.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+namespace Equinor.ProCoSys.Completion.Query.PunchItemQueries.GetPunchItemsInProject;
+
+public class GetPunchItemsInProjectQueryValidator : AbstractValidator<GetPunchItemsInProjectQuery>
+{
+    public GetPunchItemsInProjectQueryValidator(IProjectValidator projectValidator)
+    {
+        RuleLevelCascadeMode = CascadeMode.Stop;
+        ClassLevelCascadeMode = CascadeMode.Stop;
+
+        RuleFor(command => command)
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemsInProject/GetPunchItemsInProjectQueryValidator.cs:16]: command -> query?

File: src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs [hunk 3]
```diff
@@ -899,6 +1000,15 @@ public async Task GetPunchItemAttachmentDownloadUrl_AsWriter_ShouldReturnForbidd
             _punchItemGuidUnderTest,
             _attachmentGuidUnderTest,
             HttpStatusCode.Forbidden);
+
+    [TestMethod]
+    public async Task GetPunchItemAttachmentDownloadUrl_AsWriter_ShouldNotFound_WhenUnknownPunchItem()
```
Comment (jsoi-eq) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs:1005]: Nitpick: ShouldReturnNotFound

File: src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs [hunk 4]
```diff
@@ -899,6 +1000,15 @@ public async Task GetPunchItemAttachmentDownloadUrl_AsWriter_ShouldReturnForbidd
...
+        => await PunchItemsControllerTestsHelper.GetPunchItemAttachmentDownloadUrlAsync(
+            UserType.Writer,
+            TestFactory.PlantWithAccess,
+            Guid.NewGuid(), 
+            Guid.NewGuid(),
+            HttpStatusCode.NotFound);
     #endregion
```
Comment (jsoi-eq) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs:1012]: Should there be a GetPunchItemAttachmentDownloadUrl_AsWriter_ShouldNotFound_WhenUnknownAttachment test?

File: src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs [hunk 5]
```diff
@@ -954,6 +1064,17 @@ public async Task OverwriteExistingPunchItemAttachment_AsWriter_ShouldReturnForb
             TestFactory.AValidRowVersion,
             HttpStatusCode.Forbidden);
```
Comment (jsoi-eq) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs:1066]: Should we add a test for unknown file? (looks like filename is used as a kind of Id)

File: src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs [hunk 6]
```diff
@@ -1101,6 +1233,16 @@ public async Task ClearPunchItem_AsWriter_ShouldReturnForbidden_WhenNoAccessToPl
...
```
Comment (jsoi-eq) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs:1235]: Add ClearPunchItem_AsWriter_ShouldReturnBadRequest_WhenUnknownAttachment?

**Review by jsoi-eq — COMMENTED** [review]
Awaiting feedback.

File: src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs (hunk 4 above)
Comment (janid1967) [reply src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs:1012]: You got both sharp eyes and brain ;-) ... I'll fix

File: src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs (hunk 5 above)
Comment (janid1967) [reply src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs:1066]: Yes. Done

File: src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs (hunk 6 above)
Comment (jsoi-eq) [reply src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs:1258]: I can also make copy/paste errors :)
Should have been DeletePunchItemAttachment_AsWriter_ShouldReturnBadRequest_WhenUnknownAttachment

Comment (janid1967) [reply src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/PunchItems/PunchItemsControllerNegativeTests.cs:1258]: I wasn't aware of that (that you can make mistakes) ;-)
I'll add test

**Review by jsoi-eq — APPROVED** [review]

**Pull request merged**